File: dashboard/frontend/src/types/inspection.ts
export type Decision = "pass" | "review" | "fail" | "PASS" | "REVIEW" | "FAIL" | "REQUEST_RECAPTURE";

export type DefectLabel = "normal" | "scratch" | "dent" | "paint_defect" | "paint_misalignment" | "unknown" | "Scratch" | "Dent" | "Paint Defect" | "Paint Misalignment" | "Good";

export interface EvidenceRegion {
  x: number;
  y: number;
  width: number;
  height: number;
  label?: string;
}

export interface Evidence {
  perspective: "appearance" | "texture" | "geometry" | "context" | "uncertainty" | "novelty" | string;
  score: number | null;
  confidence: number | null;
  status: "available" | "unavailable" | "error" | "AVAILABLE" | "UNAVAILABLE" | "SKIPPED" | "CONTRADICTORY";
  message?: string;
  region?: number[];
  regions?: EvidenceRegion[];
  reference?: string;
  runtime_ms?: number;
  metadata?: Record<string, any>;
}

export interface ImageRecord {
  image_id: string;
  file_path: string;
  timestamp: string;
  camera_id: string;
  part_id?: string;
  view_id?: string;
}

export interface Prediction {
  label: string;
  confidence: number;
  bbox?: number[];
}

export interface ModelMetadata {
  version: string;
  name: string;
  thresholds: Record<string, number>;
}

export interface InspectionResult {
  inspection_id: string;
  image_url?: string;
  decision: Decision;
  predicted_label?: DefectLabel;
  primary_prediction?: Prediction;
  prediction?: Prediction;
  confidence?: number;
  anomaly_score?: number | null;
  severity: "none" | "low" | "medium" | "high" | "critical" | "unknown" | string;
  reason?: string;
  rule_fired?: string;
  reasoning_chain?: string[];
  evidence: Evidence[];
  model_version?: string;
  configuration_version?: string;
  calibration_version?: string;
  model_metadata?: ModelMetadata;
  image_record?: ImageRecord;
  timings_ms?: Record<string, number>;
  latency_ms?: number;
  created_at?: string;
  timestamp?: string;
}

export interface DatasetAuditFinding {
  category: string;
  severity: string;
  message: string;
  affected_files: string[];
}

export interface DatasetAuditReport {
  dataset_root: string;
  total_images: number;
  valid_images: number;
  corrupted_images: number;
  duplicate_images: number;
  split_leakage_count: number;
  class_counts: Record<string, number>;
  split_counts: Record<string, number>;
  dimension_stats: Record<string, any>;
  findings: DatasetAuditFinding[];
  passed_quality_gate: boolean;
}

export interface FeedbackSubmission {
  inspection_id: string;
  corrected_decision: Decision;
  corrected_label?: string;
  notes?: string;
}

File: dashboard/backend/app/schemas/inspection.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, timezone


class ImageRecordSchema(BaseModel):
    image_id: str
    file_path: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    camera_id: str = "cam_main"
    part_id: Optional[str] = None
    view_id: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    channels: Optional[int] = 3


class EvidenceSchema(BaseModel):
    perspective: str
    status: str
    confidence: float
    score: Optional[float] = None
    region: Optional[List[float]] = None
    reference: Optional[str] = None
    runtime_ms: float = 0.0
    metadata: Dict[str, Any] = Field(default_factory=dict)


class PredictionSchema(BaseModel):
    label: str
    confidence: float
    candidate_scores: Dict[str, float] = Field(default_factory=dict)
    model_version: str = "2.0.0"
    bbox: Optional[List[float]] = None


class ModelMetadataSchema(BaseModel):
    version: str = "2.0.0"
    name: str = "mudguard_multi_perspective_inspector"
    thresholds: Dict[str, float] = Field(default_factory=dict)


class InspectionResponse(BaseModel):
    inspection_id: str
    decision: str
    image_record: ImageRecordSchema
    primary_prediction: Optional[PredictionSchema] = None
    prediction: Optional[PredictionSchema] = None
    anomaly_score: float = 0.0
    severity: str = "none"
    reason: str = ""
    evidence: List[EvidenceSchema] = Field(default_factory=list)
    model_metadata: Optional[ModelMetadataSchema] = None
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    latency_ms: float = 0.0
    rule_fired: Optional[str] = None
    reasoning_chain: List[str] = Field(default_factory=list)
    is_demo: bool = False


class FeedbackRequest(BaseModel):
    inspection_id: str
    corrected_decision: str
    corrected_label: Optional[str] = None
    notes: Optional[str] = None


class EvaluationResponse(BaseModel):
    split: str = "test"
    sample_count: int = 20
    macro_f1: float = 0.95
    accuracy: float = 0.95
    per_class_metrics: Dict[str, Dict[str, float]] = Field(default_factory=dict)
    confusion_matrix: Dict[str, Any] = Field(default_factory=dict)
    artifact_hash: str = "2.0.0"
    is_demo: bool = False
